src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs

```diff
@@ -297,12 +298,27 @@ internal void Scan<TState>(Regex regex, string text, int textstart, ref TState s
                     if (match._matchcount[0] > 0)
                     {
                         // Hand it out to the callback in canonical form.
+                        if (!reuseMatchObject)
+                        {
+                            // We're not reusing match objects, so null out our field reference to the instance.
+                            // It'll be recreated the next time one is needed.
+                            runmatch = null;
+                        }
                         match.Tidy(runtextpos);
                         initialized = false;
                         if (!callback(ref state, match))
```

Comment:
We seem to be passing `match` to the callback here, so shouldn't we be making the local `match` object null instead of `runmatch`? (or maybe my understanding is wrong :) )

Reply:
match is the instance to be given to the callback, always non-null. We null out runmatch so that the same instance won't be handed to the callback the next time. The sequence is basically:
- If runmatch (a field) is null, create a new match instance and store it in runmatch
- store runmatch into match (a local)
- if we're not allowed to reuse runmatch again after this callback, null it out
- invoke the callback